File: leetcode/coin_change/coin_change.cpp
/*
 * Time Complexity: O(S * N)
 *   - S is the amount, N is the number of coins.
 *   - For each amount up to S, we check all N coins.
 *
 * Space Complexity: O(S)
 *   - For the dp array of size amount+1.
 */

#include <vector>
#include <algorithm>
#include <climits>

using std::vector;
using std::min;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int dp [amount+1];
        dp[0] = 0;

        sort(coins.begin(), coins.end());
        int currIndex;

        for(int c = 1; c <= amount; c++) {
            dp[c] = INT_MAX;
            for(int i : coins) {
                currIndex = c-i;
                if(currIndex < 0) break;
                else if(dp[currIndex] != INT_MAX) dp[c] = min(dp[c], dp[currIndex]+1);
            }
        }

        return dp[amount] == INT_MAX ? -1 : dp[amount];
    }
};

File: leetcode/coin_change/coin_change.java
package coin_change;

import java.util.Arrays;

class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount+1];
        dp[0] = 0; // Base case: 0 coins needed to make amount 0

        Arrays.sort(coins);

        int currIndex;

        for(int c = 1; c <= amount; c++) {
            dp[c] = Integer.MAX_VALUE;
            for(int i : coins) {
                currIndex = c-i;
                if(currIndex < 0) break;
                else if(dp[currIndex] == Integer.MAX_VALUE) continue;
                dp[c] = Math.min(dp[c], dp[currIndex]+1);
            }
        }
        return dp[amount] == Integer.MAX_VALUE ? -1 : dp[amount];
    }

    public static void main(String[] args) {
        int[] coins = {1, 2, 5};
        int amount = 11;
        Solution solution = new Solution();
        int result = solution.coinChange(coins, amount);
        System.out.println("Minimum coins needed: " + result); // Output: 3
    }
}
